lib/charms/oathkeeper/v0/auth_proxy.py

```diff
@@ -0,0 +1,448 @@
+                    f"Provided URL {url} uses http scheme. In order to make the Identity Platform work with the Proxy, run kratos in dev mode: `juju config kratos dev=True`. Don't do this in production"
+                )
+
+        # Validate headers
+        for header in self.headers:
+            if header not in ALLOWED_HEADERS:
+                raise AuthProxyConfigError(
+                    f"Unsupported header {header}, it must be one of {ALLOWED_HEADERS}"
+                )
+
+    def to_dict(self) -> Dict:
+        """Convert object to dict."""
+        return {k: v for k, v in asdict(self).items() if v is not None}
+
+
+class AuthProxyConfigChangedEvent(EventBase):
+    """Event to notify the Provider charm that the auth proxy config has changed."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        protected_urls: List[str],
+        headers: List[str],
+        allowed_endpoints: List[str],
+        relation_id: int,
+        relation_app_name: str,
+    ) -> None:
+        super().__init__(handle)
+        self.protected_urls = protected_urls
+        self.allowed_endpoints = allowed_endpoints
+        self.headers = headers
+        self.relation_id = relation_id
+        self.relation_app_name = relation_app_name
+
+    def snapshot(self) -> Dict:
+        """Save event."""
+        return {
+            "protected_urls": self.protected_urls,
+            "headers": self.headers,
+            "allowed_endpoints": self.allowed_endpoints,
+            "relation_id": self.relation_id,
+            "relation_app_name": self.relation_app_name,
+        }
+
+    def restore(self, snapshot: Dict) -> None:
+        """Restore event."""
+        self.protected_urls = snapshot["protected_urls"]
+        self.headers = snapshot["headers"]
+        self.allowed_endpoints = snapshot["allowed_endpoints"]
+        self.relation_id = snapshot["relation_id"]
+        self.relation_app_name = snapshot["relation_app_name"]
+
+    def to_auth_proxy_config(self) -> AuthProxyConfig:
+        """Convert the event information to an AuthProxyConfig object."""
+        return AuthProxyConfig(
+            self.protected_urls,
+            self.allowed_endpoints,
+            self.headers,
+        )
+
+
+class AuthProxyConfigRemovedEvent(EventBase):
+    """Event to notify the provider charm that the auth proxy config was removed."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        relation_id: int,
+    ) -> None:
+        super().__init__(handle)
+        self.relation_id = relation_id
+
+    def snapshot(self) -> Dict:
+        """Save event."""
+        return {"relation_id": self.relation_id}
+
+    def restore(self, snapshot: Dict) -> None:
+        """Restore event."""
+        self.relation_id = snapshot["relation_id"]
+
+
+class AuthProxyProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `AuthProxyProvider`."""
+
+    config_changed = EventSource(AuthProxyConfigChangedEvent)
```

Comment:
I would rename this to avoid confusion with the juju `config_changed` event.

```suggestion
    proxy_config_changed = EventSource(AuthProxyConfigChangedEvent)
```